Convert bucket aggs docs to runtime fields (#71202)
elastic/elasticsearch: docs/reference/aggregations.asciidoc

[[use-scripts-in-an-agg]]
=== Use scripts in an aggregation

> When a field doesn't exactly match the aggregation you need to do you
> should aggregate on a <<runtime,runtime field>>:

Reviewer: ```suggestion
When a field doesn't exactly match the aggregation you need, you
should aggregate on a <<runtime,runtime field>>:
```

[source,console]
----
GET /my-index-000001/_search?size=0
{
  "runtime_mappings": {
    "message.length": {
      "type": "long",
      "script": "emit(doc['message.keyword'].value.length())"
    }
  },
  "aggs": {
    "message_length": {
      "histogram": {
        "interval": 10,
        "field": "message.length"
      }
    }
  }
}
----
// TEST[setup:my_index]

////
[source,console-result]
----
{
  "timed_out": false,
  "took": "$body.took",
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0,
    "skipped": 0
  },
  "hits": "$body.hits",
  "aggregations": {
    "message_length": {
      "buckets": [
        {
          "key": 30.0,
          "doc_count": 5
        }
      ]
    }
  }
}
----
////

> Scripts will calculate the values on the fly, adding a little overhead
> to the execution of the aggretion. Some aggregations like
> <<search-aggregations-bucket-terms-aggregation,`terms`>> and
> <<search-aggregations-bucket-filters-aggregation,`filters`>> have to
> give up some optimizations.

Reviewer: ```suggestion
Scripts calculate field values dynamically, which adds overhead
to the aggregation. Certain aggregations like
<<search-aggregations-bucket-terms-aggregation,`terms`>> and
<<search-aggregations-bucket-filters-aggregation,`filters`>> are
less optimized when using scripts.
```

Reviewer: @nik9000, let me know if I totally misinterpreted your meaning here.

Author: Yeah, this is right. I think I meant something a little more nuanced with "a little" than comes across on rereading. I'll merge you suggestion and then iterate on it locally for a bit.

[[use-scripts-in-an-agg]]
=== Use scripts in an aggregation

When a field doesn't exactly match the aggregation you need, you
should aggregate on a <<runtime,runtime field>>:

[source,console]
----
GET /my-index-000001/_search?size=0
{
  "runtime_mappings": {
    "message.length": {
      "type": "long",
      "script": "emit(doc['message.keyword'].value.length())"
    }
  },
  "aggs": {
    "message_length": {
      "histogram": {
        "interval": 10,
        "field": "message.length"
      }
    }
  }
}
----
// TEST[setup:my_index]

////
[source,console-result]
----
{
  "timed_out": false,
  "took": "$body.took",
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0,
    "skipped": 0
  },
  "hits": "$body.hits",
  "aggregations": {
    "message_length": {
      "buckets": [
        {
          "key": 30.0,
          "doc_count": 5
        }
      ]
    }
  }
}
----
////

> Scripts calculate field values dynamically, which adds a little
> overhead to the aggregation. In addition to the time spent calculating,
> some aggregations like <<search-aggregations-bucket-terms-aggregation,`terms`>>
> and <<search-aggregations-bucket-filters-aggregation,`filters`>> can not use
> some of their optimizations with runtime fields. So the total performance cost
> for using a runtime field varies from aggregation to aggregation.

Reviewer: ```suggestion
Scripts calculate field values dynamically, which adds a little
overhead to the aggregation. In addition to the time spent calculating,
some aggregations like <<search-aggregations-bucket-terms-aggregation,`terms`>>
and <<search-aggregations-bucket-filters-aggregation,`filters`>> can't use
some of their optimizations with runtime fields. In total, performance costs
for using a runtime field varies from aggregation to aggregation.
```

Reviewer: I'd suggest "cannot" or "can't" (probably "can't"), and the other suggestion is just trying to lead into that last sentence a bit more.

Author: :+1:
